### PR TITLE
Convert photosReducer to TypeScript

### PR DESCRIPTION
**Pull request by jmalm**
First migration to TypeScript. This PR adds a number of types that are used for transferring photosets and photo details to frontend, and uses those in photosReducer.

Please try this out thoroughly to make sure I haven't missed any files or other aspects.

### CODE REVIEW

**Comment by jmalm** [comment]
The remaining code smells will not be changed in this PR.

**Comment by derneuere** [comment]
I get an error if I am in development mode and don't have redux tools in the browser installed:
https://stackoverflow.com/questions/57628279/composeenhancers-is-not-a-function-in-reactjs

It works once I installed the tool. I think this should be fixed as this will lead to support questions by new devs.

**Comment by derneuere** [comment]
Removing images from an album, adding images to an album or creating a new album does work, but you don't get any conformation of your action in the button right corner.

**Comment by derneuere** [comment]
The "No Timestamp" view keeps staying on loading and does not load the pictures.

**Comment by jmalm** [comment]
Ok, thanks for testing this. I will have a look.

**Comment by jmalm** [comment]
> The "No Timestamp" view keeps staying on loading and does not load the pictures.

I am not able to reproduce this (with my set of photos without timestamp?). Do you think that you could debug it and find the cause?

**Comment by jmalm** [comment]
Redux code fixed.
Album edit / create fixed.
Please test.

File: src/actions/photosActions.ts [hunk 1]
```diff
@@ -435,72 +456,25 @@ export function fetchPhotoDetail(image_hash) {
...
-          },
-        });
-      })
-      .catch(getFetchPhotosetErrorHandler(dispatch));
-  };
-}
-
+const _PaginatedPigPhotosSchema = z.object({
+  count: z.number(),
+  next: z.number().nullable(),
```
Comment (derneuere) [inline src/actions/photosActions.ts]: previous and next are strings not numbers. It returns the whole path and not just the next/previous page number. 
And I just noticed, that my count API is not needed, as it is also returned by the paginated query 😅

**Comment by derneuere** [comment]
The typing also leads to a new error, because I misused the id field 😅
https://github.com/LibrePhotos/librephotos-frontend/blob/cef7776b4d1b1011aa6a011db15dbce197b0bc03/src/layouts/photos/NoTimestampPhotosView.js#L19-L32
In this function I use the id field to do some math, but we now inferred that it is a string which leads to "11" + 1 = "111". To fix this for now we have to parse the id field as an int with parseInt(firstTempObject.id)

File: src/actions/photosActions.ts (hunk 1 above)
Comment (jmalm) [reply src/actions/photosActions.ts]: Fixed.

I didn't do anything about the count request in this PR. I think this PR serves to highlight these types of things. There is still room for improvement. :D

**Comment by jmalm** [comment]
Issues fixed. Thanks for the help! I've now added 100 random images without timestamp, so in the future, I will have these issues as well!

**Comment by derneuere** [comment]
I just checked everything again and caught something different. If you open up an album for a place, it loads indefinitely 😅 Will try to debug that tomorrow.
```{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "undefined",
      "path": [
        "id"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "received": "undefined",
      "path": [
        "title"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "array",
      "received": "undefined",
      "path": [
        "cover_photos"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "undefined",
      "path": [
        "photo_count"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "undefined",
      "path": [
        "geolocation_level"
      ],
      "message": "Required"
    }
  ],
  "name": "ZodError"
}
```

**Comment by jmalm** [comment]
Yes, I figured there would be issues with the places albums. I don't have any test data for that yet. If you care to share some few photos I can add them locally and help with the fixing.

`id` and `photo_count` being undefined seems a bit strange... I also thought that `geolocation_level` would always have a value in this response.

**Comment by derneuere** [comment]
These work well, but any photos with geolocation data should work: https://github.com/ianare/exif-samples/tree/master/jpg/gps
You also have to generate a Mapbox key for reverse geocoding the places.

File: src/actions/albumsActions.ts [hunk 2]
```diff
@@ -267,14 +334,15 @@ export function fetchPlaceAlbumsList() {
...
 }
 
-export function fetchPlaceAlbum(album_id) {
-  return function (dispatch) {
+export function fetchPlaceAlbum(album_id: string) {
+  return function (dispatch: Dispatch<any>) {
     dispatch({ type: "FETCH_PLACE_ALBUMS" });
     Server.get(`albums/place/${album_id}/`)
       .then((response) => {
+        const data = PlaceAlbumInfoSchema.parse(response.data)
```
Comment (derneuere) [inline src/actions/albumsActions.ts]: We have to add a .results to response.data.

Comment (derneuere) [reply src/actions/albumsActions.ts]: We then get the following zod error:

```
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "string",
      "path": [
        "id"
      ],
      "message": "Expected number, received string"
    },
    {
      "code": "invalid_type",
      "expected": "array",
      "received": "undefined",
      "path": [
        "cover_photos"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "undefined",
      "path": [
        "photo_count"
      ],
      "message": "Required"
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "undefined",
      "path": [
        "geolocation_level"
      ],
      "message": "Required"
    }
  ],
  "name": "ZodError"
}
```

Here is an examplary response:

```
"results": {
		"id": "14",
		"title": "Neukölln",
		"grouped_photos": [
			{
				"date": "2021-02-14",
				"location": "",
				"items": [
					{
						"id": "41e2399ff4ed78d5dc4bcbfabdd8b26f1",
						"dominantColor": "#6f829e",
						"url": "41e2399ff4ed78d5dc4bcbfabdd8b26f1",
						"location": "Wildenbruchplatz 12045 Neukölln Berlin Germany",
						"date": "2021-02-14T16:18:03+00:00",
						"birthTime": "2021-02-14 16:18:03+00:00",
						"aspectRatio": 1.33,
						"type": "image",
						"rating": 0,
						"owner": {
							"id": 1,
							"username": "admin",
							"first_name": "",
							"last_name": ""
						},
						"shared_to": []
					}
				]
			}
		]
	}
```

Should we fix the response from the backend to be more in line with the other data types?

Comment (jmalm) [reply src/actions/albumsActions.ts]: My intention was to just type them as they are, and that we can use these types to identify where we want to adjust the responses in the backend to have fewer different response types.

Comment (jmalm) [reply src/actions/albumsActions.ts]: I wonder how I could get this type so wrong...

Comment (jmalm) [reply src/actions/albumsActions.ts]: Ah, there is a serializer_class defined, but it's not used...

Comment (jmalm) [reply src/actions/albumsActions.ts]: Fixed.

**Comment by derneuere** [comment]
Found a new invalid type. If you open up a thing album, you get "No images found" found. Not sure what zod does not like about results 😅 

```
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "string",
      "path": [
        "results",
        "id"
      ],
      "message": "Expected number, received string"
    }
  ],
  "name": "ZodError"
}
```
Here is a typical response:

```
{
	"results": {
		"id": "319",
		"title": "wind farm",
		"grouped_photos": [
			{
				"date": "No timestamp",
				"location": "",
				"items": [
					{
						"id": "68b8f650361f57ed5d32923ded5389211",
						"dominantColor": "#184c8b",
						"url": "68b8f650361f57ed5d32923ded5389211",
						"location": "None",
						"date": null,
						"birthTime": "None",
						"aspectRatio": 0.56,
						"type": "image",
						"rating": 0,
						"owner": {
							"id": 1,
							"username": "admin",
							"first_name": "",
							"last_name": ""
						},
						"shared_to": []
					}
				]
			}
		]
	}
}
```

File: src/actions/albumsActions.ts (hunk 2 above)
Comment (derneuere) [reply src/actions/albumsActions.ts]: Yes, now it works 🥳

**Comment by jmalm** [comment]
Changed to string id.

It might have been that yup did some automatic type conversion, where zod requires this to be explicitly specified.

It looks like there are several id fields that are numbers coded as string. This should be fixed in the backend in the next step.

**Comment by derneuere** [comment]
I logged all the issues that were found while reviewing this PR 👍 
Looks good to me!

**Pull request merged**